### ISSUE TITLE
Build failure on Debian (kernel 5.7.0)

### ISSUE THREAD

**Issue by zeigerpuppy** [Drumfix/avb4linux#4]
I have tried building as per the instructions but unfortunately get a build failure.

Here are some details:
- OS: Debian 10 with kernel from buster-backports (actually Devuan - beowulf-backports but kernel is equivalent)
- AVB Card: HP branded Intel i210 
- kernel 5.7.0-0.bpo.2-common (from backports repo)
- gcc 8.3.0

I had to symlink my kernel headers for `autoconf.h` to be found (I think the dir has changed - see https://askubuntu.com/questions/777059/linux-autoconf-h-not-present )

```bash
cd /usr/src/linux-headers-5.7.0-0.bpo.2-amd64/include
ln -s generated ./linux
```

then onto the building:

`cmake`/`make` for `jdksavdecc-c` works fine but then I get failure with:

```bash
cd ..
make
sudo make install
```

error message:

```
cd `pwd`/kernel-module/igb
make -C /lib/modules/`uname -r`/build M=`pwd`/kernel-module/igb
make[1]: Entering directory '/usr/src/linux-headers-5.7.0-0.bpo.2-amd64'
  AR      /usr/src/avb4linux/kernel-module/igb/built-in.a
  CC [M]  /usr/src/avb4linux/kernel-module/igb/igb_main.o
/usr/src/avb4linux/kernel-module/igb/igb_main.c: In function ‘igb_io_error_detected’:
/usr/src/avb4linux/kernel-module/igb/igb_main.c:9698:3: error: implicit declaration of function ‘pci_cleanup_aer_uncorrect_error_status’ [-Werror=implicit-function-declaration]
   pci_cleanup_aer_uncorrect_error_status(pdev);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/avb4linux/kernel-module/igb/igb_main.c: In function ‘igb_set_interrupt_capability’:
/usr/src/avb4linux/kernel-module/igb/igb_main.c:1147:3: warning: this statement may fall through [-Wimplicit-fallthrough=]
   igb_reset_interrupt_capability(adapter);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/avb4linux/kernel-module/igb/igb_main.c:1148:2: note: here
  case IGB_INT_MODE_MSI:
  ^~~~
/usr/src/avb4linux/kernel-module/igb/igb_main.c: In function ‘igb_read’:
/usr/src/avb4linux/kernel-module/igb/igb_main.c:10377:2: warning: ignoring return value of ‘copy_to_user’, declared with attribute warn_unused_result [-Wunused-result]
  copy_to_user(buf, &adapter->last_event, sizeof(adapter->last_event));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/avb4linux/kernel-module/igb/igb_main.c: In function ‘__igb_notify_dca’:
/usr/src/avb4linux/kernel-module/igb/igb_main.c:6471:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
   if (dca_add_requester(dev) == E1000_SUCCESS) {
      ^
/usr/src/avb4linux/kernel-module/igb/igb_main.c:6478:2: note: here
  case DCA_PROVIDER_REMOVE:
  ^~~~
/usr/src/avb4linux/kernel-module/igb/igb_main.c: In function ‘igb_has_link’:
/usr/src/avb4linux/kernel-module/igb/igb_main.c:4796:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
   if (!hw->mac.get_link_status)
      ^
/usr/src/avb4linux/kernel-module/igb/igb_main.c:4798:2: note: here
  case e1000_media_type_internal_serdes:
  ^~~~
cc1: some warnings being treated as errors
make[3]: *** [/usr/src/linux-headers-5.7.0-0.bpo.2-common/scripts/Makefile.build:272: /usr/src/avb4linux/kernel-module/igb/igb_main.o] Error 1
make[2]: *** [/usr/src/linux-headers-5.7.0-0.bpo.2-common/Makefile:1745: /usr/src/avb4linux/kernel-module/igb] Error 2
make[1]: *** [/usr/src/linux-headers-5.7.0-0.bpo.2-common/Makefile:180: sub-make] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.7.0-0.bpo.2-amd64'
make: *** [Makefile:35: igb] Error 2
```

**Comment by zeigerpuppy** [comment]
updated note regarding symlinking to kernel headers

**Comment by zeigerpuppy** [comment]
OK, I took a bit of a stab in the dark and have managed to get it to compile on 5.7.0.
I added the following definition to end of `kernel-module/igb/kcompat.h`

```
#if ( LINUX_VERSION_CODE >= KERNEL_VERSION(5,1,0) )
#define pci_cleanup_aer_uncorrect_error_status(dev) do {} while (0)
#endif

#endif /* _KCOMPAT_H_ */
```

now to test if it works...

**Comment by zeigerpuppy** [comment]
I tried starting with `avb_up.sh` but unfortunatley it seems to fail finding the right module:

```
~$ sudo avb_up.sh eth0 48000

rmmod: ERROR: Module igb_avb is not currently loaded
Cannot get driver information: No such device
eth0: ERROR while getting interface flags: No such device
/usr/local/bin/avb_up.sh: line 19: /sys/class/net/eth0/queues/tx-0/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 20: /sys/class/net/eth0/queues/tx-1/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 21: /sys/class/net/eth0/queues/tx-2/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 22: /sys/class/net/eth0/queues/tx-3/xps_cpus: No such file or directory
eth0: ERROR while getting interface flags: No such device
eth0: ERROR while getting interface flags: No such device

...

```

after this runs, my `eth0` interface is no longer visible, I can get it back by running `sudo avb_down.sh` and `sudo ifup eth0`

my i210 network card was is interface `eth0` and is using the `igb` driver by default.  I can ping the MOTU interface and connect to the web interface.

Maybe I need to load the module somehow?

I tried:

```
~$ sudo modprobe igb_avb

modprobe: FATAL: Module igb_avb not found in directory /lib/modules/5.7.0-0.bpo.2-amd64
```

but no luck there.

p.s. I have turned off network-manager for this device, by editing `NetworkManager.conf` and adding the cards MAC address to:

```
[keyfile]
unmanaged-devices=mac:xx:xx:xx:xx:xx:xx
```

The interface is manually configured using `/etc/network/interfaces`:

```
auto eth0
iface eth0 inet static
        address 172.16.0.1
        netmask 255.255.255.0
```

**Comment by zeigerpuppy** [comment]
I have tried installing the kernel module to the kernel driver tree and loading at boot, which looks like it's working, but starting the avb service still seems to fail.

1. added driver to kernel tree

```bash
cp /root/.avb/igb_avb.ko /lib/modules/5.7.0-0.bpo.2-amd64/kernel/drivers/net/ethernet/intel/igb/igb.ko
depmod
```

2. load kernel at boot, by editing `/etc/modules`:

```
igb_avb
```

3. test that it loads

```
lsmod | grep igb
igb_avb               253952  0
snd_pcm               131072  7 snd_hda_codec_hdmi,snd_hda_intel,snd_usb_audio,igb_avb,snd_hda_codec,snd_hda_core
snd                   106496  26 snd_hda_codec_generic,snd_seq_device,snd_hda_codec_hdmi,snd_hwdep,snd_hda_intel,snd_usb_audio,snd_usbmidi_lib,igb_avb,snd_hda_codec,snd_hda_codec_realtek,snd_timer,snd_pcm,snd_rawmidi
igb                   249856  0
i2c_algo_bit           16384  2 igb,nouveau
dca                    16384  2 igb,igb_avb
ptp                    28672  3 igb,igb_avb,e1000e
```

test that it's loaded for the ethernet card:

```
lspci -s 84:00 -nnvvv
84:00.0 Ethernet controller [0200]: Intel Corporation I210 Gigabit Network Connection [8086:1533] (rev 03)
	Subsystem: Hewlett-Packard Company Ethernet I210-T1 GbE NIC [103c:0003]
	Control: I/O- Mem+ BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Interrupt: pin A routed to IRQ 45
	NUMA node: 1
	Region 0: Memory at f3200000 (32-bit, non-prefetchable) [size=1M]
	Region 3: Memory at f3300000 (32-bit, non-prefetchable) [size=16K]
	Expansion ROM at f3100000 [disabled] [size=1M]
	Capabilities: <access denied>
	Kernel modules: igb, igb_avb
```
seems all ok, to that point...

However, when I try to start avb, `sudo avb_up.sh eth0 48000`, I get the error: 

```
Cannot get driver information: No such device
eth0: ERROR while getting interface flags: No such device
/usr/local/bin/avb_up.sh: line 19: /sys/class/net/eth0/queues/tx-0/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 20: /sys/class/net/eth0/queues/tx-1/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 21: /sys/class/net/eth0/queues/tx-2/xps_cpus: No such file or directory
/usr/local/bin/avb_up.sh: line 22: /sys/class/net/eth0/queues/tx-3/xps_cpus: No such file or directory
eth0: ERROR while getting interface flags: No such device
eth0: ERROR while getting interface flags: No such device
Starting daemons on eth0
INFO     : GPTP [16:49:32:531] gPTP starting
ERROR    : GPTP [16:49:32:535] Failed to get interface address: No such device
ERROR    : GPTP [16:49:32:535] failed to initialize port
mmrp_enable failed
Error starting. Run as sudo?
error in index ioctl readingindex=0
error setting setsockopt PACKET_ADD_MEMBERSHIP
error in SIOCGIFHWADDR ioctl reading
rc = 0, mrp_register_domain
rc = 0, mrp_join_vlan
Setting samplerate to 48000.
error in sending....sendlen=-1....errno=6
Done setting samplerate.

Mac: 00-00-00-00-00-00
error in sending....sendlen=-1....errno=6
error in sending....sendlen=-1....errno=6
S+L:L=91e0f00095750000, D=2
rc = 0, done sending ready
rc = 0, advertising stream
error in sending....sendlen=-1....errno=6
...
```

also, I note the error in dmesg (MAC address hidden), `probe failed with error -22`:

```
sudo dmesg | grep igb
[    1.616358] igb: Intel(R) Gigabit Ethernet Network Driver - version 5.6.0-k
[    1.616359] igb: Copyright (c) 2007-2014 Intel Corporation.
[    1.645064] igb 0000:84:00.0: added PHC on eth0
[    1.645066] igb 0000:84:00.0: Intel(R) Gigabit Ethernet Network Connection
[    1.645068] igb 0000:84:00.0: eth0: (PCIe:2.5Gb/s:Width x1) xx:xx:xx:xx:xx:xx
[    1.645200] igb 0000:84:00.0: eth0: PBA No: G65055-011
[    1.645202] igb 0000:84:00.0: Using MSI-X interrupts. 4 rx queue(s), 4 tx queue(s)
[   25.198776] igb 0000:84:00.0 eth0: igb: eth0 NIC Link is Up 100 Mbps Full Duplex, Flow Control: RX/TX
[  278.970067] igb 0000:84:00.0: removed PHC on eth0
[  279.259672] igb_avb adapter->tx_ring_size 256
[  279.259940] igb: 0000:84:00.0: igb_check_options: VMDq option is not supported.
[  279.354585] Modules linked in: igb_avb(OE+) fuse psnap llc bnep bluetooth drbg ansi_cprng ecdh_generic ecc nfsd nfs_acl rpcsec_gss_krb5 auth_rpcgss nfsv4 dns_resolver nfs lockd grace sunrpc fscache nls_ascii nls_cp437 vfat fat parport_pc ppdev parport uinput intel_rapl_msr intel_rapl_common sb_edac x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel snd_usb_audio kvm snd_usbmidi_lib snd_rawmidi irqbypass snd_seq_device gspca_vc032x uvcvideo gspca_main videobuf2_vmalloc crc32_pclmul videobuf2_memops videobuf2_v4l2 ghash_clmulni_intel aesni_intel videobuf2_common videodev joydev nouveau libaes mc crypto_simd cryptd zfs(POE) glue_helper video rapl ttm intel_cstate drm_kms_helper intel_uncore zunicode(POE) cec drm ipmi_si zlua(POE) evdev pcspkr serio_raw zavl(POE) ipmi_devintf efi_pstore efivars ipmi_msghandler icp(POE) iTCO_wdt iTCO_vendor_support mxm_wmi watchdog snd_hda_codec_realtek snd_hda_codec_generic ledtrig_audio snd_hda_codec_hdmi snd_hda_intel snd_intel_dspcfg snd_hda_codec
[  279.354642]  ath9k snd_hda_core ath9k_common snd_hwdep ath9k_hw snd_pcm ath snd_timer mac80211 zcommon(POE) snd znvpair(POE) spl(OE) cfg80211 soundcore rfkill libarc4 mei_me mei button ext4 crc16 mbcache jbd2 crc32c_generic sr_mod cdrom ses enclosure sg sd_mod t10_pi crc_t10dif crct10dif_generic hid_generic usbhid hid isci libsas mpt3sas ahci xhci_pci libahci xhci_hcd libata ehci_pci raid_class ehci_hcd crct10dif_pclmul scsi_transport_sas crct10dif_common crc32c_intel e1000e usbcore firewire_ohci scsi_mod i2c_algo_bit dca firewire_core lpc_ich i2c_i801 mfd_core ptp crc_itu_t pps_core usb_common wmi [last unloaded: igb_avb]
[  279.354755]  igb_probe+0x9c9/0xee0 [igb_avb]
[  280.654613] igb_avb: probe of 0000:84:00.0 failed with error -22
```

some more details from `dmesg`:

```
[  278.970067] igb 0000:84:00.0: removed PHC on eth0
[  279.259220] Intel(R) Gigabit Ethernet Network Driver - version 5.3.2_AVB
[  279.259222] Copyright (c) 2007-2015 Intel Corporation.
[  279.259672] igb_avb adapter->tx_ring_size 256
[  279.259940] igb: 0000:84:00.0: igb_check_options: VMDq option is not supported.
[  279.354572] ------------[ cut here ]------------
[  279.354584] WARNING: CPU: 8 PID: 217 at net/ethtool/common.c:348 ethtool_check_ops+0x14/0x20
[  279.354585] Modules linked in: igb_avb(OE+) fuse psnap llc bnep bluetooth drbg ansi_cprng ecdh_generic ecc nfsd nfs_acl rpcsec_gss_krb5 auth_rpcgss nfsv4 dns_resolver nfs lockd grace sunrpc fscache nls_ascii nls_cp437 vfat fat parport_pc ppdev parport uinput intel_rapl_msr intel_rapl_common sb_edac x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel snd_usb_audio kvm snd_usbmidi_lib snd_rawmidi irqbypass snd_seq_device gspca_vc032x uvcvideo gspca_main videobuf2_vmalloc crc32_pclmul videobuf2_memops videobuf2_v4l2 ghash_clmulni_intel aesni_intel videobuf2_common videodev joydev nouveau libaes mc crypto_simd cryptd zfs(POE) glue_helper video rapl ttm intel_cstate drm_kms_helper intel_uncore zunicode(POE) cec drm ipmi_si zlua(POE) evdev pcspkr serio_raw zavl(POE) ipmi_devintf efi_pstore efivars ipmi_msghandler icp(POE) iTCO_wdt iTCO_vendor_support mxm_wmi watchdog snd_hda_codec_realtek snd_hda_codec_generic ledtrig_audio snd_hda_codec_hdmi snd_hda_intel snd_intel_dspcfg snd_hda_codec
[  279.354642]  ath9k snd_hda_core ath9k_common snd_hwdep ath9k_hw snd_pcm ath snd_timer mac80211 zcommon(POE) snd znvpair(POE) spl(OE) cfg80211 soundcore rfkill libarc4 mei_me mei button ext4 crc16 mbcache jbd2 crc32c_generic sr_mod cdrom ses enclosure sg sd_mod t10_pi crc_t10dif crct10dif_generic hid_generic usbhid hid isci libsas mpt3sas ahci xhci_pci libahci xhci_hcd libata ehci_pci raid_class ehci_hcd crct10dif_pclmul scsi_transport_sas crct10dif_common crc32c_intel e1000e usbcore firewire_ohci scsi_mod i2c_algo_bit dca firewire_core lpc_ich i2c_i801 mfd_core ptp crc_itu_t pps_core usb_common wmi [last unloaded: igb_avb]
[  279.354685] CPU: 8 PID: 217 Comm: kworker/8:1 Tainted: P           OE     5.7.0-0.bpo.2-amd64 #1 Debian 5.7.10-1~bpo10+1
[  279.354686] Hardware name: XXXXXXXXX
[  279.354694] Workqueue: events work_for_cpu_fn
[  279.354698] RIP: 0010:ethtool_check_ops+0x14/0x20
[  279.354701] Code: ee 41 bc f4 ff ff ff eb e6 66 66 2e 0f 1f 84 00 00 00 00 00 66 90 66 66 66 66 90 31 c0 48 83 7f 70 00 74 0d 8b 17 85 d2 75 07 <0f> 0b b8 ea ff ff ff c3 0f 1f 40 00 66 66 66 66 90 48 8b 97 98 08
[  279.354702] RSP: 0018:ffffaefcc976fd90 EFLAGS: 00010246
[  279.354704] RAX: 0000000000000000 RBX: ffff9b3edac8c000 RCX: 0000000000000000
[  279.354705] RDX: 0000000000000000 RSI: 0000000000000018 RDI: ffffffffc16c3c40
[  279.354707] RBP: ffff9b3ede8640b0 R08: 0000dcf6bb5fc020 R09: 0000000000000008
[  279.354708] R10: 0000000000000000 R11: ffffffff84050640 R12: ffffffff840f14c0
[  279.354709] R13: ffff9b3edac8d91c R14: ffff9b3edac8cd50 R15: ffff9b3edac8c000
[  279.354711] FS:  0000000000000000(0000) GS:ffff9b3edfa00000(0000) knlGS:0000000000000000
[  279.354712] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  279.354714] CR2: 00005557c0f23020 CR3: 00000001d880a002 CR4: 00000000000606e0
[  279.354715] Call Trace:
[  279.354724]  register_netdevice+0x69/0x570
[  279.354730]  register_netdev+0x1a/0x30
[  279.354755]  igb_probe+0x9c9/0xee0 [igb_avb]
[  279.354763]  ? __switch_to_asm+0x40/0x70
[  279.354769]  local_pci_probe+0x42/0x80
[  279.354772]  work_for_cpu_fn+0x16/0x20
[  279.354776]  process_one_work+0x1ad/0x370
[  279.354778]  worker_thread+0x1cf/0x390
[  279.354781]  ? create_worker+0x1a0/0x1a0
[  279.354784]  kthread+0x112/0x130
[  279.354787]  ? kthread_park+0x80/0x80
[  279.354790]  ret_from_fork+0x35/0x40
[  279.354794] ---[ end trace 271366aacb54106b ]---
[  280.654613] igb_avb: probe of 0000:84:00.0 failed with error -22
```

after these errors, the interface is unusable. but can be brought back up by:

```bash
modprobe igb
```

**Comment by Drumfix** [comment]
Closed because we moved to 5.10 already
